Source/ui_unix/InputBindingManager.cpp

```diff
@@ -0,0 +1,577 @@
+{
+	std::string prefBase = Framework::CConfig::MakePreferenceName(buttonBase, CONFIG_POVHATBINDING_PREFIX);
+	config.SetPreferenceString(Framework::CConfig::MakePreferenceName(prefBase, CONFIG_BINDINGINFO_DEVICE).c_str(), string_format("%02x:%02x:%02x:%02x:%02x:%02x", m_binding.device[0], m_binding.device[1], m_binding.device[2], m_binding.device[3], m_binding.device[4], m_binding.device[5]).c_str());
+	config.SetPreferenceInteger(Framework::CConfig::MakePreferenceName(prefBase, CONFIG_BINDINGINFO_ID).c_str(), m_binding.id);
+	config.SetPreferenceInteger(Framework::CConfig::MakePreferenceName(prefBase, CONFIG_BINDINGINFO_TYPE).c_str(), m_binding.type);
+	config.SetPreferenceInteger(Framework::CConfig::MakePreferenceName(prefBase, CONFIG_POVHATBINDING_REFVALUE).c_str(), m_refValue);
+}
+
+void CInputBindingManager::CPovHatBinding::Load(Framework::CConfig& config, const char* buttonBase)
+{
+	std::string prefBase = Framework::CConfig::MakePreferenceName(buttonBase, CONFIG_POVHATBINDING_PREFIX);
+	CGamePadDeviceListener::ParseMAC(m_binding.device, config.GetPreferenceString(Framework::CConfig::MakePreferenceName(prefBase, CONFIG_BINDINGINFO_DEVICE).c_str()));
+	m_binding.id = config.GetPreferenceInteger(Framework::CConfig::MakePreferenceName(prefBase, CONFIG_BINDINGINFO_ID).c_str());
+	m_binding.type = config.GetPreferenceInteger(Framework::CConfig::MakePreferenceName(prefBase, CONFIG_BINDINGINFO_TYPE).c_str());
+	m_refValue = config.GetPreferenceInteger(Framework::CConfig::MakePreferenceName(prefBase, CONFIG_POVHATBINDING_REFVALUE).c_str());
+}
+
+void CInputBindingManager::CPovHatBinding::RegisterPreferences(Framework::CConfig& config, const char* buttonBase)
+{
+	std::string prefBase = Framework::CConfig::MakePreferenceName(buttonBase, CONFIG_POVHATBINDING_PREFIX);
+	config.RegisterPreferenceString(Framework::CConfig::MakePreferenceName(prefBase, CONFIG_BINDINGINFO_DEVICE).c_str(), "00:00:00:00:00");
+	config.RegisterPreferenceInteger(Framework::CConfig::MakePreferenceName(prefBase, CONFIG_BINDINGINFO_ID).c_str(), 0);
+	config.RegisterPreferenceInteger(Framework::CConfig::MakePreferenceName(prefBase, CONFIG_BINDINGINFO_TYPE).c_str(), 0);
+	config.RegisterPreferenceInteger(Framework::CConfig::MakePreferenceName(prefBase, CONFIG_POVHATBINDING_REFVALUE).c_str(), -1);
+}
+
+void CInputBindingManager::CPovHatBinding::ProcessEvent(std::array<uint32, 6> device, uint32 id, uint32 value)
+{
+	if(id != m_binding.id) return;
+	if(device != m_binding.device) return;
+	m_value = value;
+}
+
+std::string CInputBindingManager::CPovHatBinding::GetDescription() const
+{
+	if(m_binding.device == std::array<uint32, 6>{0})
+	{
+		return QString("Key: %1").arg(QKeySequence(m_binding.id).toString()).toStdString();
+	}
+	else
+	{
+		return QString("Key: %1").arg(libevdev_event_code_get_name(m_binding.type, m_binding.id)).toStdString();
+	}
+}
+
+uint32 CInputBindingManager::CPovHatBinding::GetValue() const
+{
+	if(m_value != m_refValue) return 0;
```

Comment:
@jpd002 quick comment/question about this.


in the Win32 implementation, you had the line `if(m_value == -1) return 0;`.
however, this didn't work in my testing with a PS4 controller, a POVHAT (in this case the D-Pad was reported as such) returned a value of `-1,0,1` along a single axis (Left/Right & Up/Down), so returning 0 meant that Left & Up did nothing.